ledger/streaming/client/src/backfill.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A BlockStream that backfills missing indices from a given BlockFetcher.
+
+use displaydoc::Display;
+use futures::{FutureExt, Stream, StreamExt};
+use mc_common::logger::{log, Logger};
+use mc_ledger_streaming_api::{BlockFetcher, BlockStream, BlockStreamComponents, Result};
+use mc_transaction_core::BlockIndex;
+use std::{pin::Pin, sync::Arc};
+
+/// A [BlockStream] that backfills another [BlockStream] using a [BlockFetcher].
+#[derive(Debug, Display)]
+pub struct BackfillingStream<S: BlockStream + 'static, F: BlockFetcher + 'static> {
+    upstream: S,
+    fetcher: Arc<F>,
+    logger: Logger,
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BackfillingStream<S, F> {
+    /// Instantiate a [BackfillingStream].
+    pub fn new(upstream: S, fetcher: F, logger: Logger) -> Self {
+        Self {
+            upstream,
+            fetcher: Arc::new(fetcher),
+            logger,
+        }
+    }
+}
+
+impl<S: BlockStream + 'static, F: BlockFetcher + 'static> BlockStream for BackfillingStream<S, F> {
+    type Stream = impl Stream<Item = Result<BlockStreamComponents>>;
+
+    fn get_block_stream(&self, starting_height: u64) -> Result<Self::Stream> {
+        self.upstream
+            .get_block_stream(starting_height)
+            .map(|upstream| {
+                backfill_stream(
+                    upstream,
+                    starting_height,
+                    self.fetcher.clone(),
+                    self.logger.clone(),
+                )
+            })
+    }
+}
+
+fn backfill_stream<
+    S: Stream<Item = Result<BlockStreamComponents>> + 'static,
+    F: BlockFetcher + 'static,
+>(
+    upstream: S,
+    starting_height: u64,
+    fetcher: Arc<F>,
+    logger: Logger,
+) -> impl Stream<Item = Result<BlockStreamComponents>> {
+    use futures::stream::{empty, once};
+
+    let mut prev_index: Option<BlockIndex> = None;
+    upstream.flat_map(
+        move |result| -> Pin<Box<dyn Stream<Item = Result<BlockStreamComponents>>>> {
```

Comment:
No, I had to box here because of the different stream types across the match arms. This boxed Stream is consumed by the `flat_map` call above.

I agree we should profile these pipelines when we have all the pieces.